### ISSUE TITLE
non-seeded random numbers in tests

### ISSUE THREAD

**Issue by jeromedockes** [nilearn/nilearn#2534]
these parts of the unit tests seem to use a random number generator with no
fixed seed, which can cause sporadic test failures (e.g.
https://api.travis-ci.com/v3/job/357101203/log.txt):

```
./image/tests/test_resampling.py:47:    data = np.random.randint(0, 10, shape)
./image/tests/test_resampling.py:223:    data = np.random.randint(0, 10, shape)
./image/tests/test_resampling.py:556:    source_affine[:, 3] = np.append(np.random.randint(0, 4, 3), 1)
./image/tests/test_resampling.py:625:    for theta, phi in np.random.randint(4, size=(5, 2)):
./image/tests/test_resampling.py:731:    x = np.random.random((10,))
./image/tests/test_resampling.py:732:    y = np.random.random((10,))
./image/tests/test_resampling.py:733:    z = np.random.random((10,))
./image/tests/test_image.py:341:    data = np.random.randn(4, 5, 7)
./image/tests/test_image.py:497:    fake_fmri_data = np.random.rand(10, 10, 10, 10)
./image/tests/test_image.py:498:    fake_affine = np.random.rand(4, 4)
./glm/tests/test_contrasts.py:31:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:41:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:53:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:64:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:93:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:111:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:123:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_contrasts.py:142:    X1, V1 = np.random.randn(p), np.ones(p)
./glm/tests/test_second_level.py:100:        X = np.random.randn(p, q)
./glm/tests/test_second_level.py:177:        X = np.random.randn(p, q)
./glm/tests/test_second_level.py:347:        X = pd.DataFrame(np.random.rand(4, 2), columns=['r1', 'r2'])
./glm/tests/test_second_level.py:392:        X = pd.DataFrame(np.random.rand(4, 2), columns=['r1', 'r2'])
./glm/tests/test_first_level.py:221:    X, Y = np.random.randn(p, q), np.random.randn(p, n)
./glm/tests/test_first_level.py:248:    u = np.random.randn(*shape)
./glm/tests/test_first_level.py:249:    mean = 100 * np.random.rand(shape[1]) + 1
./glm/tests/test_thresholding.py:19:    np.random.shuffle(x)
./glm/tests/test_dmtx.py:59:    durations = 5 + 5 * np.random.rand(len(onsets))
./glm/tests/test_dmtx.py:60:    values = 1 + np.random.rand(len(onsets))
./glm/tests/test_dmtx.py:72:    values = 1 + np.random.rand(len(onsets))
./glm/tests/test_dmtx.py:116:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:121:    ax = np.random.randn(127, 4)
./glm/tests/test_dmtx.py:127:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:141:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:345:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:359:    ax = np.random.randn(128, 4)
./glm/tests/test_dmtx.py:420:    ax = np.random.randn(128, 4)
./glm/tests/test_hemodynamic_models.py:83:    X = np.random.randn(100, 5)
./glm/tests/test_hemodynamic_models.py:92:    X = np.random.randn(100)
./glm/tests/test_utils.py:23:    X = np.random.randn(n, p)
./glm/tests/test_utils.py:33:    p = np.random.rand(10)
./glm/tests/test_utils.py:43:    x = np.random.rand(n) / n
./glm/tests/test_utils.py:44:    A = np.random.rand(n, n) / n
./glm/tests/test_utils.py:52:    x = np.random.randn(n, 3)
./glm/tests/test_utils.py:55:        A = np.random.randn(120, n)
./glm/tests/test_utils.py:58:    i = np.random.randint(3)
./glm/tests/test_utils.py:77:    X = np.random.randn(shape[0], shape[1], shape[2])
./glm/tests/test_paradigm.py:27:    durations = 5 + 5 * np.random.rand(len(onsets))
./glm/tests/test_paradigm.py:28:    values = np.random.rand(len(onsets))
./glm/tests/test_paradigm.py:40:    values = np.random.rand(len(onsets))
./regions/tests/test_region_extractor.py:69:    img = np.zeros((30, 30, 30)) + 0.1 * np.random.randn(30, 30, 30)
./regions/tests/test_region_extractor.py:92:    map_img = np.zeros((30, 30, 30)) + 0.1 * np.random.randn(30, 30, 30)
./plotting/tests/test_img_plotting.py:211:    data = np.random.randn(10, 10, 10)
./plotting/tests/test_img_plotting.py:432:    maps_img = nibabel.Nifti1Image(np.random.random((10, 10, 10)), np.eye(4))
./plotting/tests/test_img_plotting.py:1299:    adjacency_matrix = np.random.random((4, 4))
./input_data/tests/test_nifti_spheres_masker.py:12:    data = np.random.random((3, 3, 3, 5))
./input_data/tests/test_nifti_spheres_masker.py:23:    data = np.random.random((3, 3, 3, 5))
./input_data/tests/test_nifti_spheres_masker.py:48:    data = np.random.random((3, 3, 3, 5))
./input_data/tests/test_nifti_spheres_masker.py:85:    data = np.random.random(shape + (5,))
./input_data/tests/test_nifti_spheres_masker.py:110:    data = np.random.random(shape)
./input_data/tests/test_nifti_spheres_masker.py:141:    data_without_nans = np.random.random((9, 9, 9))
./input_data/tests/test_nifti_spheres_masker.py:161:    data = np.random.random((3, 3, 3, 5))
./input_data/tests/test_nifti_masker.py:185:    data_5d = [np.random.random((10, 10, 10, 1)) for i in range(5)]
./input_data/tests/test_nifti_masker.py:206:    data_5d = [np.random.random((10, 10, 10, 3)) for i in range(5)]
./input_data/tests/test_nifti_masker.py:320:    img = np.random.rand(9, 9, 5)
./input_data/tests/test_nifti_masker.py:400:    signals = np.random.randn(np.prod(data_shape), n_samples)
./input_data/tests/test_nifti_masker.py:401:    means = np.random.randn(np.prod(data_shape), 1) * 50 + 1000
./input_data/tests/test_nifti_maps_masker.py:310:    signals = np.random.randn(np.prod(data_shape), n_samples)
./input_data/tests/test_nifti_maps_masker.py:311:    means = np.random.randn(np.prod(data_shape), 1) * 50 + 1000
./input_data/tests/test_nifti_labels_masker.py:323:    signals = np.random.randn(np.prod(data_shape), n_samples)
./input_data/tests/test_nifti_labels_masker.py:324:    means = np.random.randn(np.prod(data_shape), 1) * 50 + 1000
./input_data/tests/test_multi_nifti_masker.py:166:    imgs = [Nifti1Image(np.random.rand(9, 9, 5), np.eye(4)),
./input_data/tests/test_multi_nifti_masker.py:167:            Nifti1Image(np.random.rand(9, 9, 5), np.eye(4))]
./input_data/tests/test_multi_nifti_masker.py:202:    signals = np.random.randn(2, np.prod(data_shape), n_samples)
./input_data/tests/test_multi_nifti_masker.py:203:    means = np.random.randn(2, np.prod(data_shape), 1) * 50 + 1000
./tests/test_niimg_conversions.py:492:        buffer.append(Nifti1Image(np.random.random((10, 10, 10)), np.eye(4)))
./tests/test_niimg_conversions.py:498:    l = [Nifti1Image(np.random.random((10, 10, 10)), np.eye(4))
./tests/test_masking.py:116:    img1 = Nifti1Image(np.full((9, 9, 9), np.random.rand()), np.eye(4))
./tests/test_masking.py:117:    img2 = Nifti1Image(np.full((9, 9, 9), np.random.rand()), np.eye(4))
./tests/test_masking.py:432:    imgs1 = [Nifti1Image(np.random.randn(9, 9, 9), np.eye(4)),
./tests/test_masking.py:433:             Nifti1Image(np.random.randn(9, 9, 9), np.eye(4))]
./tests/test_masking.py:436:    imgs2 = [Nifti1Image(np.random.randn(9, 9, 9), np.eye(4)),
./tests/test_masking.py:437:             Nifti1Image(np.random.randn(9, 9, 9), np.eye(4))]
./tests/test_niimg.py:63:            dataobj = np.random.uniform(0, 255,
./tests/test_segmentation.py:14:    img = np.zeros((30, 30, 30)) + 0.1 * np.random.randn(30, 30, 30)
./tests/test_segmentation.py:47:    np.random.seed(42)
./tests/test_segmentation.py:48:    img = np.random.normal(size=(3, 3, 3, 3, 3))
./tests/test_segmentation.py:54:    img = np.random.normal(size=(10, 10))
./tests/test_segmentation.py:67:    data = np.zeros((5, 5)) + 0.1 * np.random.randn(5, 5)
./decoding/tests/test_searchlight.py:83:    groups = np.random.permutation(np.arange(frames, dtype=int) >
```

**Comment by chouhanaryan** [comment]
Hi, I'd like to take this up. From what I can gather, this involves setting a seed for all of the mentioned code and that should be it. Let me know if I'm missing something.

**Comment by jeromedockes** [comment]
> Hi, I'd like to take this up. From what I can gather, this involves setting a seed for all of the mentioned code and that should be it. Let me know if I'm missing something.

Thanks! that's right, replacing calls of functions such as
`np.random.randn` by methods of a seeded random number generator eg
`np.random.RandomState(0).randn`

**Comment by chouhanaryan** [comment]
Hi, thanks for your reply! So I've been going through the codebase and I spot a couple of inconsistencies in the way random number generation is handled in different places.

One of the following is done in most places -
1. No seed is set - https://github.com/nilearn/nilearn/blob/master/nilearn/image/tests/test_resampling.py#L47
2. sklearn's `sklearn.utils.check_random_state` function is used to generate a RandomState object -
https://github.com/nilearn/nilearn/blob/master/nilearn/mass_univariate/tests/test_permuted_least_squares.py#L96
3. Seed is set using `np.random.RandomState` in the function itself - https://github.com/nilearn/nilearn/blob/master/nilearn/image/tests/test_resampling.py#L83
4. A local variable passed in as an argument to the test function is used to set the RandomState (different from 3) - https://github.com/nilearn/nilearn/blob/master/nilearn/decoding/tests/test_fista.py#L26
5. `np.random.seed()` is used - https://github.com/nilearn/nilearn/blob/master/nilearn/tests/test_segmentation.py#L47 (this is the only place where this is done)

No. 2 seems to be the most common so I'm going to go ahead and start using that for the required changes.
However, I'm curious whether an inconsistency like this is important enough to fix or not since it doesn't have a significant impact on neither readability nor performance. What are your thoughts?

**Comment by bthirion** [comment]
I think that it's good to fix it.
But this is not a blocker, this is why it has been in the current state for some years.

**Comment by jeromedockes** [comment]
Aryan Chouhan <notifications@github.com> writes:

> Hi, thanks for your reply! So I've been going through the codebase and I spot a couple of inconsistencies in the way random number generation is handled in different places.
>
> One of the following is done in most places -
> 1. No seed is set - https://github.com/nilearn/nilearn/blob/master/nilearn/image/tests/test_resampling.py#L47

this is what we want to remove

> 2. sklearn's `sklearn.utils.check_random_state` function is used to generate a RandomState object -
> https://github.com/nilearn/nilearn/blob/master/nilearn/mass_univariate/tests/test_permuted_least_squares.py#L96
> 3. Seed is set using `np.random.RandomState` in the function itself - https://github.com/nilearn/nilearn/blob/master/nilearn/image/tests/test_resampling.py#L83
> 4. A local variable passed in as an argument to the test function is used to set the RandomState (different from 3) - https://github.com/nilearn/nilearn/blob/master/nilearn/decoding/tests/test_fista.py#L26

these (2, 3, and 4) are basically the same thing. passing the seed as an
argument would only be useful if used in combination with
`pytest.mark.parametrize` to run the test function several times with
different random values -- can be useful for example to catch some
numerical stability issues that don't show up systematically. In example
from the link above it doesn't need to be an argument.

> 5. `np.random.seed()` is used - https://github.com/nilearn/nilearn/blob/master/nilearn/tests/test_segmentation.py#L47 (this is the only place where this is done)

this sets a global seed; it is a legacy function and if we can using a
RandomState (or the newer Generator) would be better

For new functions and tests, we should now get a `Generator` by calling
`np.random.default_rng(seed)` instead:

https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.default_rng

>
> No. 2 seems to be the most common so I'm going to go ahead and start using that for the required changes.
> However, I'm curious whether an inconsistency like this is important enough to fix or not since it doesn't have a significant impact on neither readability nor performance. What are your thoughts?
>

tests that use a `RandomState` are fine and there is no need to change
them. the other ones (those that don't set a seed for sure, and possibly
those that use `np.random.seed` as well) should be modified to use a
`Generator` with a fixed seed.

**Comment by jeromedockes** [comment]
> tests that use a `RandomState` are fine and there is no need to change
> them. the other ones (those that don't set a seed for sure, and possibly
> those that use `np.random.seed` as well) should be modified to use a
> `Generator` with a fixed seed.

also if you have already started working on the PR and added
`RandomState` to some tests that's perfectly fine you don't need to
change these to use `get_default_rng`.

**Comment by thomasbazeille** [comment]
if we don't want new tests to come without a seed, we should also add a mention of what should be done in the contributing documentation

**Comment by chouhanaryan** [comment]
So I went ahead and made the [changes ](https://github.com/nilearn/nilearn/compare/master...chouhanaryan:rng-tests-2534) to the following modules - images, glm, plotting and regions.

I _might've_ actually gone a bit overboard because I switched out `RandomState` with `Generator` at quite a few places. But it wasn't until much later that I realized my [builds](https://travis-ci.com/github/chouhanaryan/nilearn/builds/186881349) were failing because the first job uses Numpy 1.11 but `numpy.random.Generator` was added only in [1.17](https://numpy.org/doc/stable/release/1.17.0-notes.html#new-extensible-numpy-random-module-with-selectable-random-number-generators). All of the other jobs are passing.

How do I proceed? 😅 Do we want to use `Generator` or should I shelve it for now and just use `RandomState`?

**Comment by jeromedockes** [comment]
> So I went ahead and made the [changes ](https://github.com/nilearn/nilearn/compare/master...chouhanaryan:rng-tests-2534) to the following modules - images, glm, plotting and regions.
>
> I _might've_ actually gone a bit overboard because I switched out `RandomState` with `Generator` at quite a few places. But it wasn't until much later that I realized my [builds](https://travis-ci.com/github/chouhanaryan/nilearn/builds/186881349) were failing because the first job uses Numpy 1.11 but `numpy.random.Generator` was added only in [1.17](https://numpy.org/doc/stable/release/1.17.0-notes.html#new-extensible-numpy-random-module-with-selectable-random-number-generators). All of the other jobs are passing.
>
> How do I proceed? 😅 Do we want to use `Generator` or should I shelve it for now and just use `RandomState`? 

Ah yes, sorry about that I didn't realize it was so recent. Then we have
to keep using RandomState for now.

Fortunately it should be easy to change using a simple textual search
and replace "np.random.default_rng" by "np.random.RandomState" -- please
let us know if you meet some difficulties

As you already got started, I suggest that you open a pull request (you
can start its title with "[WIP]" to indicate it is still work in
progress), it will be easier to continue this discussion there and
comment on the diff.
